agent/src/main/java/com/teamscale/jacoco/agent/AgentBase.java

```diff
@@ -45,18 +60,65 @@ public AgentBase(AgentOptions options) throws IllegalStateException {
 
 		logger.info("Starting JaCoCo agent for process {} with options: {}",
 				ManagementFactory.getRuntimeMXBean().getName(), getOptionsObjectToLog());
+		retryUnsuccessfulUploads(options);
 		if (options.getHttpServerPort() != null) {
 			try {
 				initServer();
 			} catch (Exception e) {
-				throw new IllegalStateException(
-						"Control server not started.", e);
+				throw new IllegalStateException("Control server not started.", e);
 			}
 		}
 	}
 
 	/**
-	 * Lazily generated string representation of the command line arguments to print to the log.
+	 * If we have coverage that was leftover because of previously unsuccessful
+	 * coverage uploads, we retry to upload them again with the same configuration
+	 * as in the previous try.
+	 *
+	 * This method expects that the .properties file of the leftover coverage is
+	 * constructed in a specific order.
+	 * 
+	 */
+	private void retryUnsuccessfulUploads(AgentOptions options) {
+		if (options.getOutputDirectory() == null) {
+			return;
+		}
+		List<File> reuploadCandidates = FileSystemUtils.listFilesRecursively(
+				options.getOutputDirectory().getParent().toFile(),
+				filepath -> filepath.getName().contains(RETRY_UPLOAD_FILE_SUFFIX));
+		for (File file : reuploadCandidates) {
+			try {
+				logger.info("Retrying previously unsuccessful coverage upload for file {}.", file);
+				String[] config = FileSystemUtils.readFile(file).split("\n");
+				TeamscaleServer server = new TeamscaleServer();
+				server.url = HttpUrl.parse(StringUtils.stripPrefix(config[0], "url="));
+				server.project = StringUtils.stripPrefix(config[1], "project=");
+				server.userName = StringUtils.stripPrefix(config[2], "userName=");
+				server.userAccessToken = StringUtils.stripPrefix(config[3], "userAccessToken=");
+				server.partition = StringUtils.stripPrefix(config[4], "partition=");
+				server.commit = CommitDescriptor.parse(StringUtils.stripPrefix(config[5], "commit="));
+				String revision = StringUtils.stripPrefix(config[6], "revision=");
+				if (revision.equals("null")) {
+					revision = null;
+				}
+				server.revision = revision;
+				server.setMessage(StringUtils.stripPrefix(config[7], "message=").replace("$nl$", "\n"));
```

Comment:
You could extract this into a new method to fix the finding and to make the code a bit cleaner